File: src/app/modules/page/dashboard/core/dashboard.service.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ApiService } from '@core/api/api.service';
import { environment } from '@env';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Dashboard } from '../model/dashboard.model';

@Injectable()
export class DashboardService {
  constructor(private apiService: ApiService) {}

  public getStats(): Observable<Dashboard> {
    return this.apiService.post(`/kpi-stats`).pipe(map((payload: { data: Dashboard; status: number }) => payload.data));
  }

  public getGeolocationOfCountry(iso: string): Observable<any> {
    const httpParams = new HttpParams()
      .set('access_token', environment.mapBoxAccessToken)
      .set('language', 'en-EN')
      .set('types', 'country');
    return this.apiService.getBy(`https://api.mapbox.com/geocoding/v5/mapbox.places/${iso}.json`, httpParams);
  }
}

File: src/main/java/net/catenax/traceability/assets/domain/service/DashboardService.java
/********************************************************************************
 * Copyright (c) 2021,2022 Contributors to the CatenaX (ng) GitHub Organisation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/

package net.catenax.traceability.assets.domain.service;

import net.catenax.traceability.assets.domain.model.Dashboard;
import net.catenax.traceability.assets.domain.ports.AssetRepository;
import net.catenax.traceability.common.security.JwtAuthentication;
import net.catenax.traceability.common.security.JwtRole;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Component;

@Component
public class DashboardService {

	private final AssetRepository assetRepository;

	public DashboardService(AssetRepository assetRepository) {
		this.assetRepository = assetRepository;
	}

	public Dashboard getDashboard(JwtAuthentication jwtAuthentication) {
		long totalAssets = assetRepository.countAssets();
		long myParts = assetRepository.countMyAssets();
		long pendingInvestigations = assetRepository.countPendingInvestigations();

		// first check if user has admin or supervisor role
		if (jwtAuthentication.hasAtLeastOneRole(JwtRole.ADMIN, JwtRole.SUPERVISOR)) {
			return new Dashboard(myParts, totalAssets - myParts, pendingInvestigations);
		}
		else if (jwtAuthentication.hasRole(JwtRole.USER)) {
			return new Dashboard(myParts, null, pendingInvestigations);
		}else {
			throw new AccessDeniedException("User has invalid role to access the dashboard.");
		}
	}
}
